### ISSUE TITLE
Add alternative serialization formats for built in managers

### ISSUE THREAD

**Issue by mkleinbort-ic** [dagster-io/dagster#13592]
### What's the use case?

I'm very happy with the pickle IO Mangers, but in practice I find myself implementing my own managers because (in my case) parquet files are much better than pickle.

Looking at https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-azure/dagster_azure/adls2/io_manager.py I see that the use of pickle is almost entirely abitrary - it's only used twice

`(Line 95)         obj = pickle.loads(stream.readall())`
`(Line 117)         pickled_obj = pickle.dumps(obj, PICKLE_PROTOCOL)`

Could you add a more flexible version of the pickle io manager? One that allows me to replace `pickle` with anything that implements a pickle-like interface - specifically `.loads()` and `.dumps()` eg. `dill`.
 of pickle.

### Ideas of implementation

If `pickle` above were an IO manager attribute, it would enable some very cool things. For example:

```python
class PolarsParquetWriter:
    def __init__(self, read_kwargs=None, write_kwargs=None):
        self.read_kwargs = read_kwargs if read_kwargs else dict()
        self.write_kwargs= write_kwargsif write_kwargs else dict()

    def loads(self, data)->pl.DataFrame:
         rerturn pl.read_parquet(data, **self.read_kwargs)

    def dumps(self, obj:pl.DataFrame)->io.BytesIO:
        buffer = io.BytesIO()
        obj.write_parquet(buffer **self.write_kwargs)
        buffer.seek(0)
        return buffer
```

The io-manager could then be initialized to use an instance of `PolarsParquetWriter` instead

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by clairelin135** [comment]
Any thoughts on this @jamiedemaria ?

**Comment by jamiedemaria** [comment]
hey @mkleinbort-ic  have you taken a look at the [UPathIOManager](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/storage/upath_io_manager.py)? It sounds pretty similar to what you're requesting. Basically it's a base class that implements most of the boiler-plate logic (getting paths, figuring out partitions etc) and to write a new io manager you just implement a sub-class that can take an object and write it to a specified path (dump_to_path) and read an object from a path (load_from_path). The [FSIOManager](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/storage/fs_io_manager.py) is a implementation of the UPathIOManager  if you'd like to see an example implementation to see if it fits your needs

**Comment by Christakou** [comment]
Hi @mkleinbort-ic I dont know if you're still having this issue, but I have just implemented an IO manager that sounds like it's exactly what you need: 
`import io
import pdb
from typing import Any, Dict, Optional, Union
import pandas as pd

from dagster import (
    ConfigurableIOManager,
    InputContext,
    MetadataValue,
    OutputContext,
    ResourceDependency,
    _check as check,
)
from dagster._core.storage.upath_io_manager import UPathIOManager
from upath import UPath
from dagster_aws.s3 import S3Resource
from pydantic import Field
from dagster._utils.cached_method import cached_method






class S3PandasParquetIOInternalManager(UPathIOManager):
    def __init__(
        self,
        s3_bucket: str,
        s3_session: Any,
        s3_prefix: Optional[str] = None,
    ):
        self.bucket = check.str_param(s3_bucket, "s3_bucket")
        check.opt_str_param(s3_prefix, "s3_prefix")
        self.s3 = s3_session
        self.s3.list_objects(Bucket=s3_bucket, Prefix=s3_prefix, MaxKeys=1)
        base_path = UPath(s3_prefix) if s3_prefix else None
        super().__init__(base_path=base_path)


        
    def _get_path(self, context: InputContext):
        return super()._get_path(context).with_suffix(".parquet")



    def load_from_path(self, context: InputContext, path: UPath) -> pd.DataFrame:
        try:
            
            s3_obj = io.BytesIO(self.s3.get_object(Bucket=self.bucket, Key=str(path))["Body"].read())
            return pd.read_parquet(s3_obj)
        except self.s3.exceptions.NoSuchKey:
            raise FileNotFoundError(f"Could not find file {path} in S3 bucket {self.bucket}")

    def dump_to_path(self, context: OutputContext, obj: pd.DataFrame, path: UPath) -> None:
        if self.path_exists(path):
            context.log.warning(f"Removing existing S3 object: {path}")
            self.unlink(path)

        
        parquet_buffer = io.BytesIO()
        obj.to_parquet(parquet_buffer, engine="pyarrow")
        parquet_buffer.seek(0) # can use put_object instead of upload_fileobj? 
        self.s3.upload_fileobj(parquet_buffer, self.bucket, str(path))

    def path_exists(self, path: UPath) -> bool:
        try:
            self.s3.get_object(Bucket=self.bucket, Key=str(path))
        except self.s3.exceptions.NoSuchKey:
            return False
        return True

    def get_loading_input_log_message(self, path: UPath) -> str:
        return f"Loading S3 object from: {self._uri_for_path(path)}"

    def get_writing_output_log_message(self, path: UPath) -> str:
        return f"Writing S3 object at: {self._uri_for_path(path)}"

    def unlink(self, path: UPath) -> None:
        self.s3.delete_object(Bucket=self.bucket, Key=str(path))

    def make_directory(self, path: UPath) -> None:
        # It is not necessary to create directories in S3
        return None

    def get_metadata(self, context: OutputContext, obj: Any) -> Dict[str, MetadataValue]:
        path = self._get_path(context)
        return {"uri": MetadataValue.path(self._uri_for_path(path))}

    def get_op_output_relative_path(self, context: Union[InputContext, OutputContext]) -> UPath:
        return UPath("storage", super().get_op_output_relative_path(context))

    def _uri_for_path(self, path: UPath) -> str:
        return f"s3://{self.bucket}/{path}"
    
    
    
    

class S3PandasParquetIOManager(ConfigurableIOManager):

    s3_resource: ResourceDependency[S3Resource]
    s3_bucket: str = Field(description="S3 bucket to use for the file manager.")
    s3_prefix: str = Field(
        default="dagster", description="Prefix to use for the S3 bucket for this file manager."
    )

    @classmethod
    def _is_dagster_maintained(cls) -> bool:
        return False

    @cached_method
    def inner_io_manager(self) -> S3PandasParquetIOInternalManager:
        return S3PandasParquetIOInternalManager(
            s3_bucket=self.s3_bucket,
            s3_session=self.s3_resource.get_client(),
            s3_prefix=self.s3_prefix,
        )

    def load_input(self, context: InputContext) -> Any:
        return self.inner_io_manager().load_input(context)

    def handle_output(self, context: OutputContext, obj: Any) -> None:
        return self.inner_io_manager().handle_output(context, obj)
`

@jamiedemaria would this be something we can add to ship with dagster?

**Comment by mkleinbort-ic** [comment]
The code looks good - very similar to what I ended up implementing for Azure. I might combine the two - thank you for the input. 

I think what my ask was asking for (not very clearly) was for some dependency inversion functionality for IO managers. We can inherit from `UPathIOManager` but it'd be nice to just parametrize the actual loading/writing of the file object. 

You can imagine adding two parameters to the `__init__` for `reader`/`writer`. Then replacing `pickle.loads` / `pickle.dumps` with a generic `reader`/`writer` which are callables that either take a filepath like object and return the asset or store the asset in a filepath like object.

Eg.

```python 

# To use dill
reader = dill.loads
writer = dill.dumps 

# To use pandas with csv file
reader = pd.read_csv
writer = pd.to_csv

# To use polars parquet
reader = pl.read_parquet
writer = pl.write_parquet
```

A lot of these readers/writers could be pre-configured via `functools.partial`

In short - it'd be nice to have an IOManager that can be configured to different reader/writers without using inheritance.

**Comment by Christakou** [comment]
Agreed @mkleinbort-ic would be nice to have the ability to inject any object  implementing a read/write interface

**Comment by the-xentropy** [comment]
I agree it would be nice to offer easier 'swapping' of backing system and format for trivial cases but I'd like to highlight that there's a world of complexity here if we wanted to have a truly generic interface for portability. Not every file-system is POSIX-like, and not every format is file-like. Toy example: Can an IO file manager implement partial read caching based on extended filesystem file attributes?

Having swappables for cases which are "resource-like" (e.g. data can be stably read/written as units), lock-free and "posix-like" for simpler cases would be awesome though - but I think such an interface and implementation contracts would have to be cautiously and defensively designed to avoid implicit dependencies between them and make expectations clear to engineers. For the toy example above for example - read caching could break formats that need to write metadata to disk as things are being read (e.g. locks).

So, big +1 but assumptions about the format and backing storage would have to be very clearly documented to avoid creating more issues than it solves.

**Comment by mkleinbort-ic** [comment]
Any views on whether this will be integrated into Dagster? I could copy the code and define my own IO manager, but it's always nice to benefit from a centrally maintained implementation.

**Comment by jamiedemaria** [comment]
I'd say it's pretty unlikely we will do something like this in the near-term given our other priorities

**Comment by mkleinbort-ic** [comment]
Thank you - knowing the work won't be scrap in 1.6.2 is good enough reason to go ahead for me

**Comment by mkleinbort-ic** [comment]
While I'm at it - I see my old implementation was:

1) Inheriting from `IOManager`

And

2) defining `handle_output` / `load_input`

This works well, but can anyone confirm that the current best practice is to 

1) inherit from `UPathIOManager`
2) define `dump_to_path` & `load_from_path`

**Comment by jamiedemaria** [comment]
You can do either! `UPathIOManager` is specifically for file based storage systems (the local filesystem, GCS, S3, etc). It handles a lot of computation around partitions for you, which is why it might be useful if you're in a filesystem based storage system. But if you're not, or you just prefer using `IOManager`, that is still considered a best practice too

**Comment by sryza** [comment]
Closing this because of the existence of UPathIOManager
